File: checkmate-archunit/src/main/java/io/github/malczuuu/checkmate/archunit/NamingRules.java
package io.github.malczuuu.checkmate.archunit;

import static com.tngtech.archunit.base.DescribedPredicate.describe;
import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;

import com.tngtech.archunit.lang.ArchRule;

/** Provides shared ArchUnit rules for test class naming conventions. */
public final class NamingRules {

  /**
   * This verifies that all test classes (containing {@code @Test} or {@code @ArchTest}) must end
   * with {@code "...Tests"} suffix, and not {@code "...Test"}.
   */
  public static final ArchRule TEST_CLASSES_MUST_BE_PLURAL =
      classes()
          .that()
          .containAnyMethodsThat(
              describe(
                  "are annotated with @Test or @ArchTest",
                  m ->
                      m.isAnnotatedWith(org.junit.jupiter.api.Test.class)
                          || m.isAnnotatedWith(com.tngtech.archunit.junit.ArchTest.class)))
          .and()
          .areNotNestedClasses()
          .should()
          .haveSimpleNameEndingWith("Tests");

  private NamingRules() {}
}

File: checkmate-archunit/src/main/kotlin/io/github/malczuuu/checkmate/archunit/NamingRules.kt
package io.github.malczuuu.checkmate.archunit

import com.tngtech.archunit.base.DescribedPredicate
import com.tngtech.archunit.junit.ArchTest
import com.tngtech.archunit.lang.ArchRule
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition
import org.junit.jupiter.api.Test

/**
 * Shared ArchUnit rules enforcing naming conventions across all modules in the Checkmate project.
 *
 * Import this object in any module's arch-test suite to apply the same constraints uniformly.
 */
object NamingRules {

  /**
   * An [ArchRule] that asserts every class containing at least one method annotated with [Test] or
   * [ArchTest] has a simple name ending with `"Tests"` (plural form) rather than `"Test"`.
   */
  val TEST_CLASSES_MUST_BE_PLURAL: ArchRule =
      ArchRuleDefinition.classes()
          .that()
          .containAnyMethodsThat(
              DescribedPredicate.describe("are annotated with @Test or @ArchTest") {
                it.isAnnotatedWith(Test::class.java) || it.isAnnotatedWith(ArchTest::class.java)
              }
          )
          .should()
          .haveSimpleNameEndingWith("Tests")
}
